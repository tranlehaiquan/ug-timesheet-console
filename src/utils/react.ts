import * as React from 'react'

type IdentityFunction = <T>(fn: T) => T

export const typedMemo: IdentityFunction = React.memo

export const getDisplayName = <T extends any = any>(WrappedComponent: React.ComponentType<T>): string => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
