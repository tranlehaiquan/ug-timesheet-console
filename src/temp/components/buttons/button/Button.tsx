import * as React from 'react'
import classNames from 'classnames'

import { LoadingSpinner } from '../../loader/spinner/LoadingSpinner'

export type ButtonTypes = 'primary' | 'secondary' | 'transparent'

type IButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & React.RefAttributes<HTMLButtonElement> & {
  /**
  * The type of button to be displayed. Primary | Secondary | Transparent
  */
  buttonType: ButtonTypes,
  /**
  * Pass additional class names
  */
  className?: string,
  /**
  * Disable the button
  */
  disabled?: boolean
  /**
  * Compact sized buttonâ€”used when space is tight - allowing more room for content.
  */
  compact?: boolean
  /**
  * Loading button state
  */
  loading?: boolean
}

/**
* Button component
*/
export const Button: React.RefForwardingComponent<HTMLButtonElement, IButtonProps> = React.forwardRef<HTMLButtonElement, IButtonProps>(({ buttonType, compact, loading, children, className, ...otherProps }, ref) => {

  const buttonClassMap: { [key in ButtonTypes]: string } = {
    primary: 'sked-button--primary',
    secondary: 'sked-button--secondary',
    transparent: 'sked-button--transparent'
  }

  const classes = classNames(
    'sked-button',
    buttonClassMap[buttonType],
    {
      'sked-button--compact': compact,
      'sk-pointer-events-none': loading
    },
    className
  )

  return (
    <button { ...otherProps }  className={ classes } ref={ ref }>
      <div className={ loading ? 'sk-invisible' : 'sk-visible' }>
        { children }
      </div>
      <div className="sk-pin sk-absolute sk-m-auto sk-flex sk-justify-center sk-items-center">
        <LoadingSpinner className={ loading ? 'sk-visible' : 'sk-invisible' } size={ 18 } />
      </div>
    </button>
  )
})
