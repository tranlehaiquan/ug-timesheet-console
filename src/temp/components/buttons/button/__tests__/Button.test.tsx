import * as React from "react";
import { render } from "enzyme";

import { Button } from "../Button";

describe("Buttons", () => {
  test("it renders primary correctly", () => {
    // Arrange
    const component = render(<Button buttonType="primary" />);

    // Assert
    expect(component).toMatchSnapshot();
  });

  test("it renders secondary correctly", () => {
    // Arrange
    const component = render(<Button buttonType="secondary" />);

    // Assert
    expect(component).toMatchSnapshot();
  });

  test("it renders transparent correctly", () => {
    // Arrange
    const component = render(<Button buttonType="transparent" />);

    // Assert
    expect(component).toMatchSnapshot();
  });

  test("it renders primary - compact correctly", () => {
    // Arrange
    const component = render(<Button buttonType="primary" compact />);

    // Assert
    expect(component).toMatchSnapshot();
  });

  test("it renders primary - disabled correctly", () => {
    // Arrange
    const component = render(<Button buttonType="primary" disabled />);

    // Assert
    expect(component).toMatchSnapshot();
  });

  test("it renders primary - loading correctly", () => {
    // Arrange
    const component = render(<Button buttonType="primary" loading />);

    // Assert
    expect(component).toMatchSnapshot();
  });
});
