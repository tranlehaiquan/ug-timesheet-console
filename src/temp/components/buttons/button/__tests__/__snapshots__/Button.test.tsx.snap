// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons it renders primary - compact correctly 1`] = `
<button
  class="sked-button sked-button--primary sked-button--compact"
>
  <div
    class="sk-visible"
  />
  <div
    class="sk-pin sk-absolute sk-m-auto sk-flex sk-justify-center sk-items-center"
  >
    <div
      class="loading-spinner sk-invisible"
    >
      <svg
        class="loading-spinner--circular"
        height="18"
        viewBox="25 25 50 50"
        width="18"
      >
        <circle
          class="loading-spinner--path"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke="currentColor"
          stroke-miterlimit="10"
          stroke-width="5"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Buttons it renders primary - disabled correctly 1`] = `
<button
  class="sked-button sked-button--primary"
  disabled=""
>
  <div
    class="sk-visible"
  />
  <div
    class="sk-pin sk-absolute sk-m-auto sk-flex sk-justify-center sk-items-center"
  >
    <div
      class="loading-spinner sk-invisible"
    >
      <svg
        class="loading-spinner--circular"
        height="18"
        viewBox="25 25 50 50"
        width="18"
      >
        <circle
          class="loading-spinner--path"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke="currentColor"
          stroke-miterlimit="10"
          stroke-width="5"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Buttons it renders primary - loading correctly 1`] = `
<button
  class="sked-button sked-button--primary sk-pointer-events-none"
>
  <div
    class="sk-invisible"
  />
  <div
    class="sk-pin sk-absolute sk-m-auto sk-flex sk-justify-center sk-items-center"
  >
    <div
      class="loading-spinner sk-visible"
    >
      <svg
        class="loading-spinner--circular"
        height="18"
        viewBox="25 25 50 50"
        width="18"
      >
        <circle
          class="loading-spinner--path"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke="currentColor"
          stroke-miterlimit="10"
          stroke-width="5"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Buttons it renders primary correctly 1`] = `
<button
  class="sked-button sked-button--primary"
>
  <div
    class="sk-visible"
  />
  <div
    class="sk-pin sk-absolute sk-m-auto sk-flex sk-justify-center sk-items-center"
  >
    <div
      class="loading-spinner sk-invisible"
    >
      <svg
        class="loading-spinner--circular"
        height="18"
        viewBox="25 25 50 50"
        width="18"
      >
        <circle
          class="loading-spinner--path"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke="currentColor"
          stroke-miterlimit="10"
          stroke-width="5"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Buttons it renders secondary correctly 1`] = `
<button
  class="sked-button sked-button--secondary"
>
  <div
    class="sk-visible"
  />
  <div
    class="sk-pin sk-absolute sk-m-auto sk-flex sk-justify-center sk-items-center"
  >
    <div
      class="loading-spinner sk-invisible"
    >
      <svg
        class="loading-spinner--circular"
        height="18"
        viewBox="25 25 50 50"
        width="18"
      >
        <circle
          class="loading-spinner--path"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke="currentColor"
          stroke-miterlimit="10"
          stroke-width="5"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Buttons it renders transparent correctly 1`] = `
<button
  class="sked-button sked-button--transparent"
>
  <div
    class="sk-visible"
  />
  <div
    class="sk-pin sk-absolute sk-m-auto sk-flex sk-justify-center sk-items-center"
  >
    <div
      class="loading-spinner sk-invisible"
    >
      <svg
        class="loading-spinner--circular"
        height="18"
        viewBox="25 25 50 50"
        width="18"
      >
        <circle
          class="loading-spinner--path"
          cx="50"
          cy="50"
          fill="none"
          r="20"
          stroke="currentColor"
          stroke-miterlimit="10"
          stroke-width="5"
        />
      </svg>
    </div>
  </div>
</button>
`;
